block/validate.go

```diff
@@ -0,0 +1,89 @@
+package block
+
+import (
+	"context"
+	"errors"
+
+	"github.com/dymensionxyz/dymint/settlement"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+	"github.com/tendermint/tendermint/libs/pubsub"
+)
+
+// onNewStateUpdateFinalized will update the last validated height with the last finalized height
+func (m *Manager) onNewStateUpdateFinalized(event pubsub.Message) {
+	eventData, ok := event.Data().(*settlement.EventDataNewBatch)
+	if !ok {
+		m.logger.Error("onNewStateUpdateFinalized", "err", "wrong event data received")
+		return
+	}
+	m.UpdateLastValidatedHeight(eventData.EndHeight)
+}
+
+// ValidateLoop listens for syncing events (from new state update or from initial syncing) and validates state updates to the last submitted height.
+func (m *Manager) ValidateLoop(ctx context.Context) error {
+	lastValidatedHeight, err := m.Store.LoadValidationHeight()
+	if err != nil {
+		m.logger.Debug("validation height not loaded", "err", err)
+	}
+	m.UpdateLastValidatedHeight(lastValidatedHeight)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-m.validateC:
+
+			m.logger.Info("validating state updates to target height", "targetHeight", m.LastSubmittedHeight.Load())
+
+			for currH := m.NextValidationHeight(); currH <= m.LastSubmittedHeight.Load(); currH = m.NextValidationHeight() {
+
+				// get next batch that needs to be validated from SL
+				batch, err := m.SLClient.GetBatchAtHeight(currH)
+				if err != nil {
+					m.logger.Error("failed batch retrieval", "error", err)
+					continue
```

Comment:
i see it will retry if fails to get  a batch (e.g. network issue) inside the function. so yes you are right. i changed to stop here and i added unhealthy event